src/starkware_utils/commitment_tree/patricia_tree/virtual_calculation_node.rs

```diff
@@ -0,0 +1,650 @@
+use std::any::Any;
+use std::marker::PhantomData;
+
+use num_bigint::BigUint;
+
+use crate::starkware_utils::commitment_tree::base_types::{Height, Length, NodePath};
+use crate::starkware_utils::commitment_tree::binary_fact_tree::BinaryFactDict;
+use crate::starkware_utils::commitment_tree::binary_fact_tree_node::read_node_fact;
+use crate::starkware_utils::commitment_tree::calculation::{
+    Calculation, CalculationNode, ConstantCalculation, DependencyWrapper, HashCalculation, LeafFactCalculation,
+    NodeFactDict,
+};
+use crate::starkware_utils::commitment_tree::errors::{CombineError, TreeError};
+use crate::starkware_utils::commitment_tree::leaf_fact::LeafFact;
+use crate::starkware_utils::commitment_tree::patricia_tree::nodes::{
+    verify_path_value, BinaryNodeFact, EdgeNodeFact, PatriciaNodeFact,
+};
+use crate::starkware_utils::commitment_tree::patricia_tree::patricia_tree::EMPTY_NODE_HASH;
+use crate::starkware_utils::commitment_tree::patricia_tree::virtual_patricia_node::VirtualPatriciaNode;
+use crate::storage::storage::{Fact, FactFetchingContext, HashFunctionType, Storage, StorageError};
+
+#[derive(Debug, PartialEq)]
+pub struct BinaryCalculation<S, H, LF>
+where
+    S: Storage + 'static,
+    H: HashFunctionType + 'static,
+    LF: LeafFact<S, H>,
+{
+    left: Box<HashCalculationImpl<S, H, LF>>,
+    right: Box<HashCalculationImpl<S, H, LF>>,
+    _s: PhantomData<S>,
+    _h: PhantomData<H>,
```

Comment:
You can use a tuple: `_phantom: PhantomData<(S, H)>`

Reply:
Thanks, didn't think about that!